tools/cli/adminDBTimers.go

```diff
@@ -0,0 +1,140 @@
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package cli
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/gocql/gocql"
+	"github.com/urfave/cli"
+
+	"github.com/uber/cadence/common/log/loggerimpl"
+	"github.com/uber/cadence/common/persistence"
+	cassp "github.com/uber/cadence/common/persistence/cassandra"
+)
+
+// AdminTimers is used to list scheduled timers.
+func AdminTimers(c *cli.Context) {
+	lowerShardBound := c.Int(FlagLowerShardBound)
+	upperShardBound := c.Int(FlagUpperShardBound)
+	concurrency := c.Int(FlagConcurrency)
+	batchSize := c.Int(FlagBatchSize)
+	startDate := c.String(FlagStartDate)
+	endDate := c.String(FlagEndDate)
+	domainId := c.String(FlagDomainID)
+	skipErrMode := c.Bool(FlagSkipErrorMode)
+
+	session := connectToCassandra(c)
+	defer session.Close()
+
+	ctx, cancel := newContextForLongPoll(c)
+	defer cancel()
+
+	st, err := parseSingleTs(startDate)
+	if err != nil {
+		ErrorAndExit("wrong date format for "+FlagStartDate, err)
+	}
+	et, err := parseSingleTs(endDate)
+	if err != nil {
+		ErrorAndExit("wrong date format for "+FlagEndDate, err)
+	}
+
+	group, ctx := errgroup.WithContext(ctx)
+	throttle := make(chan struct{}, concurrency)
+	for shardID := lowerShardBound; shardID < upperShardBound; shardID++ {
+		shardID := shardID
+		group.Go(func() error {
+			throttle <- struct{}{}
+			defer func() {
+				<-throttle
+			}()
+
+			return printTimers(session, shardID, batchSize, st, et, domainId, skipErrMode)
+		})
+	}
+
+	if err := group.Wait(); err != nil {
+		ErrorAndExit("getting timers", err)
+	}
+}
+
+func printTimers(
+	session *gocql.Session,
+	shardID, batchSize int,
+	startDate, endDate time.Time,
+	domainId string,
+	skipErrMode bool,
+) error {
+	execStore, err := cassp.NewWorkflowExecutionPersistence(shardID, session, loggerimpl.NewNopLogger())
+
+	if err != nil {
+		return errors.Wrapf(err, "cannot get execution manager for shardId %d", shardID)
+	}
+
+	var token []byte
+	isFirstIteration := true
+	for isFirstIteration || len(token) != 0 {
+		isFirstIteration = false
+		req := persistence.GetTimerIndexTasksRequest{
+			MinTimestamp:  startDate,
+			MaxTimestamp:  endDate,
+			BatchSize:     batchSize,
+			NextPageToken: token,
+		}
+		resp, err := execStore.GetTimerIndexTasks(&req)
+
+		if err != nil {
+			return errors.Wrapf(err, "cannot get timer tasks for shardId %d", shardID)
+		}
+
+		token = resp.NextPageToken
+		if len(resp.Timers) < 1 {
```

Comment:
This is unnecessary, i.e. the for loop below will be skipped anyway if Timers array is empty?

Reply:
good point